iroh/src/magicsock/relay_actor.rs

```diff
@@ -274,27 +279,52 @@ impl ActiveRelayActor {
     async fn run(mut self) -> anyhow::Result<()> {
         inc!(MagicsockMetrics, num_relay_conns_added);
 
-        loop {
-            let Some(client) = self.run_dialing().instrument(info_span!("dialing")).await else {
-                break;
-            };
-            match self
-                .run_connected(client)
-                .instrument(info_span!("connected"))
-                .await
-            {
-                Ok(_) => break,
-                Err(err) => {
-                    debug!("Connection to relay server lost: {err:#}");
-                    continue;
-                }
+        let mut backoff = Self::build_backoff();
+
+        while let Err(err) = self.run_once().await {
+            warn!("Relay connection failed: {err:#}");
+            if self.pong_received {
+                // If the relay connection remained established long enough so that we received a pong
+                // from the relay server, we reset the backoff and attempt to reconnect immediately.
+                self.pong_received = false;
+                backoff = Self::build_backoff();
```

Comment:
Ugh, there are not even docs for the `ExponentialBackoff` returned by this.  And the trait does not allow resetting.  Personally I find it strange that you can't reset, but whatever.  Maybe I'm the weird one.

Reply:
Yeah, the `ExponentialBackoff` is `doc_hidden`, I was confused as well. This is why I opted for returning an `impl Backoff` to not have to use a doc_hidden type in the signature. And I went to look for a reset method at first, too... so you're not the only weird one here ;)